lxd/cluster/heartbeat.go

```diff
@@ -70,22 +71,41 @@ func Heartbeat(gateway *Gateway, cluster *db.Cluster) (task.Func, task.Schedule)
 			logger.Warnf("Failed to get current cluster nodes: %v", err)
 			return
 		}
+
 		heartbeats := make([]time.Time, len(nodes))
+		heartbeatsLock := sync.Mutex{}
```

Comment:
I don't think a lock is needed, since the heartbeats slice is created upfront with the correct size and each go routine will modify a different slot in the slice.

Reply:
Yeah, wasn't completely sure about that one so went with the safe approach :)

If this is documented to be safe, I'll drop the lock.

Reply:
Yeah, I'm pretty sure the lock is not needed: the [memory model](https://golang.org/ref/mem) says that 

> Programs that modify data being simultaneously accessed by multiple goroutines must serialize such access.

and in this case the data is not being accessed simultaneously (the slice header does not change, and each goroutine writes a different slice slot, without reading any other slot).

 Running ```go run -race main.go``` against the program below confirms that:

```golang
package main

import (
	"fmt"
	"sync"
	"time"
)

func main() {
	heartbeats := make([]time.Time, 1000)
	heartbeatsWg := sync.WaitGroup{}
	for i := range heartbeats {
		heartbeatsWg.Add(1)
		defer heartbeatsWg.Done()
		go func(i int) {
			heartbeats[i] = time.Now()
		}(i)
	}
	heartbeatsWg.Wait()
	fmt.Println("Done")
}
```